### ISSUE TITLE
CLO LogCollector/LogForwarding for longterm storage

### ISSUE THREAD

**Issue by tumido** [operate-first/apps#232]
https://github.com/operate-first/apps/pull/230 will limit the stored logs to 7 days. DeCorus wants to analyze logs from much longer periods of time. We need to figure out a scalable storage and forward the logs in there.

/sig devops

**Comment by tumido** [comment]
Related to https://github.com/operate-first/support/issues/83

**Comment by HumairAK** [comment]
>from much longer periods of time

Do we know how much longer?

**Comment by tumido** [comment]
yes, the requirement is to have at least half a year worth of data

**Comment by drbwa** [comment]
And we can even put 12 or 24 months of raw logs to good use. 

Compressed JSON files or maybe parquet (with a bit of investment to understand the log schema and making sure it saves significant space over compressed JSON files) in object storage might be an option.

**Comment by drbwa** [comment]
I have looked into using the [ES Snapshot](https://www.elastic.co/guide/en/elasticsearch/reference/7.11/snapshot-restore.html) feature to back up ES indices to object storage as a potential solution to long-term storage of logs on MOC.

So far, I have come to the conclusion that this is not a suitable approach for our needs. It seems easy to set up and has some nice features (e.g., filter on snapshot and restore, existing integration with S3). However, in order to then use historic data, we would need to restore into a similar ES cluster (replicas, shards), which means that a) we need twice the resources to host such a cluster and b) can still only afford to restore and process seven days of data at a time (using a similarly configured cluster). Seems like too much of a headache.

It might be better to just store the logs in object storage to enable clients to download and transform as they see fit.

Two options:

* We could look into using the [fluentd S3 plugin](https://docs.fluentd.org/output/s3) to write logs to Ceph in addition to ES.
* We could write a custom script that runs hourly and reads the latest data from ES, compresses it and writes it to a bucket in Ceph.

How much space could we get in Ceph for logs?

Looking at a recent snapshot of `infra` and `app` indices, it looks the size of these indices with seven days of data is around 300 GB (in total). We could test how well this compresses using different approaches and would then want to store a multiple of that in object storage.

What do you think @HumairAK @tumido ?

/cc @ronensc

**Comment by drbwa** [comment]
And of course, not to forget [Loki](https://github.com/grafana/loki) might be an option as well, as mentioned in <https://github.com/open-infrastructure-labs/telemetrywg/issues/4>.

**Comment by 4n4nd** [comment]
I think Loki might already have a solution for storage, I can look into it and then we can figure out how to get CLO to push logs to Loki instead of elasticsearch

**Comment by HumairAK** [comment]
I'm not sure how fluentd plugins work, a brief skim over looks like it's going to require us to mess with the fluentd config, which I can't see a way to do using the CLO custom resource. We could set the CLO to an unmanaged state and change it, but this looks like it's [unsupported](https://docs.openshift.com/container-platform/4.6/logging/config/cluster-logging-maintenance-support.html#cluster-logging-maintenance-support-list_cluster-logging-unsupported). 

Looking at the CLO docs, they seem to offer an abstraction called `ClusterLogForwarder`, see [here](https://docs.openshift.com/container-platform/4.6/logging/cluster-logging-external.html#cluster-logging-collector-log-forwarding-about_cluster-logging-external) more info. Seems like `kafka` is listed as an option there, could we maybe forward to our kafka (provisioend via odh) instance and use an s3 connector to send those logs to a bucket?

**Comment by drbwa** [comment]
As far as I understand, Loki indexes only metadata about logs and stores the raw logs on object storage in compressed form. It promises to achieve much less storage overhead than a db like Elasticsearch that relies on much larger indices. You can also stage storage across tiers to, for example, have freshest sets of logs on a StatefulSets, then move older ones out to S3 and eventually to some form of archival/cold storage. So far, Loki sounds great.

I just don't understand the impact on queryability (how do you get things out at the other end) at the moment, but will look into that.

**Comment by drbwa** [comment]
@4n4nd @HumairAK Does the presence of this [OperateFirst Readme on Loki](https://www.operate-first.cloud/users/apps/docs/observatorium/loki/README.md) indicate that there is an existing deployment of Loki that we could 'borrow' for the purposes of long-term storage of logs via Loki?

**Comment by tumido** [comment]
Yes, we have loki deployed. As well as the rest of the Observatorium stack. We plan on using the Observatorium to be used in parallel to CLO stack, so we have both options available for any research. 

https://github.com/operate-first/apps/blob/master/observatorium/base/instance/observatorium.yaml

**Comment by 4n4nd** [comment]
> we could 'borrow' for the purposes of long-term storage of logs via Loki

yess definitely, that is one of the intentions for our Loki deployment. Please lmk if there's any other information that you might need.

**Comment by 4n4nd** [comment]
I was looking into pushing logs from CLO to Loki instead of Elasticsearch, but I couldn't find much information about it. I am trying to discuss it with the CLO team and it looks like they are working on it. All I could find was this PR (https://github.com/openshift/cluster-logging-operator/pull/428).

I think as an alternative solution, we can look into what @HumairAK suggested about forwarding logs from CLO to kafka and then we could set up a consumer to push these logs into loki.

**Comment by 4n4nd** [comment]
I asked the Cluster Logging team and they said using Kafka and a consumer sink to loki seems like a reasonable solution until CLO adds this feature ([slack discussion](https://coreos.slack.com/archives/CB3HXM2QK/p1617041685220600)).

Also, @HumairAK found a project called Vector (https://github.com/timberio/vector/) which we might be able to use to source logs from kafka and sink them to Loki.

**Comment by 4n4nd** [comment]
@HumairAK @tumido sending application logs to Loki has been working quite well, should I start sending Cluster logs as well as Infra logs to Loki?

**Comment by HumairAK** [comment]
I'm good with that -- can we also just bring down ES/Kibana, and use Loki as the one source of logs? Not sure if we have the resources/man-power to maintain both.

**Comment by tumido** [comment]
@4n4nd 
> should I start sending Cluster logs as well as Infra logs to Loki?
Let's do that! :+1: 

@HumairAK 
> can we also just bring down ES/Kibana
Let's keep it running, I don't see a reason why to remove it now. If it breaks we can solve it as a low priority task, we don't have to pay that much attention to it.

**Comment by 4n4nd** [comment]
We have long term storage of logs now available. Logs are being stored in Loki.
